guides/unit-testing.md

```diff
@@ -0,0 +1,973 @@
+
+## Colocate test files with implementation files
+
+Test files are much easier to find when they sit next to the code they test.
+
+🚫
+
+```
+src/
+  permission-controller.ts
+test/
+  permission-controller.ts
+```
+
+✅
+
+```
+src/
+  permission-controller.ts
+  permission-controller.test.ts
+```
+
+## Use `describe` to group tests for the module under test
+
+Wrap all tests for a file in a `describe` block named the same as the file itself:
+
+🚫
+
+``` typescript
+// file-utils.ts
+
+export function isFile() {
+  // ...
+}
+
+export function isDirectory() {
+  // ...
+}
+
+// file-utils.test.ts
+
+describe('isFile', () => {
+  // ...
+});
+
+describe('isDirectory', () => {
+  // ...
+});
+```
+
+✅
+
+``` typescript
+// file-utils.ts
+
+export function isFile() {
+  // ...
+}
+
+export function isDirectory() {
+  // ...
+}
+
+// file-utils.test.ts
+
+describe('file-utils', () => {
+  describe('isFile', () => {
+    // ...
+  });
+
+  describe('isDirectory', () => {
+    // ...
+  });
+});
+```
+
+## Use `describe` to group tests for methods
+
+Using `describe` to wrap tests for a method makes it easier to spot these tests in a large file and makes it possible to run them on their own (using `.only`). It also helps to establish the subject of the test (the "it") and therefore keeps the test well named and focused.
+
+🚫
+
+``` typescript
+describe('KeyringController', () => {
+  it('adds a new account to the given keyring', () => {
+    // ...
+  });
+
+  it('removes the account identified by the given address from its associated keyring', () => {
+    // ...
+  });
+});
+```
+
+✅
+
+``` typescript
+describe('KeyringController', () => {
+  describe('addAccount', () => {
+    it('adds a new account to the given keyring', () => {
+      // ...
+    });
+  });
+
+  describe('removeAccount', () => {
+    it('removes the account identified by the given address from its associated keyring', () => {
+      // ...
+    });
+  });
+});
+```
+
+## 💡 Use `describe` to group tests for scenarios
+
+If you have many tests that verify the behavior of a piece of code under a particular condition, you might find it helpful to wrap those tests with a `describe` block. Use a phrase starting with `if ...` or `when ...` so as to form a sentence when combined with the test name.
+
+1️⃣
+
+``` typescript
+it('calls addToken on the tokens controller if "addToken" is true', () => {
+  // ...
+})
+
+it('adds the new token to the state if "addToken" is true', () => {
+  // ...
+})
+
+it('returns a promise that resolves to true if "addToken" is true', () => {
+  // ...
+})
+```
+
+2️⃣
+
+``` typescript
+describe('if "addToken" is true', () => {
+  it('calls addToken on the tokens controller', () => {
+    // ...
+  });
+
+  it('adds the new token to the state', () => {
+    // ...
+  });
+
+  it('returns a promise that resolves to true', () => {
+    // ...
+  });
+});
+```
+
+## Use `it` to specify known behavior for a unit of code
```

Comment:
Originally I had a section which explained the purpose of test names and why they're arranged using `describe` and `it`. Essentially, you can draw a straight line from Jest to the birth of behavior-driven development. However, BDD is a subject unto itself, and I didn't want to overload people with information.